### PR TITLE
fix #8979 docs: add nimbus web glean application structure adr

### LINKED ISSUE #8979: Write Glean application structure ADR
https://mozilla-hub.atlassian.net/browse/EXP-3575

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-3583)

### PR DESCRIPTION
**Pull request by jeddai**
Because

- Cirrus (Nimbus for the web) necessitates a deeper discussion and decision regarding the structuring of Glean applications,

This commit

- Adds an ADR addressing those concerns.

---

[Rendered version](https://github.com/jeddai/experimenter/blob/nimbus-web-glean-structure-adr/docs/adrs/0011-nimbus-web-glean-application-structure.md)

### CODE REVIEW

**Comment by danielkberry** [comment]
Looked over the analysis and the options seem correct. At this point, my preference is 2) `One Glean app ID per product domain`. It has the most advantages, and least disadvantages. Regarding the downsides to that option (namely splitting out a "deploy app" like `cirrus_pocket` from a "target app" like `pocket_ios`), I feel like we'd have to make those changes anyway.

File: docs/adrs/0011-nimbus-web-glean-application-structure.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+* Bad, because data retention cannot be handled separately for each application or product domain.
+* Bad, because it breaks away from our existing Glean usage practices.
+* Bad, because it requires changes to Jetstream.
+  * At this time, the nature and timeline of these changes is unclear.
+* Bad, because it could require more changes to Experimenter than we already expect.
+  * While we will still need to add new applications, it is also possible we will need to adjust how applications are handled in Experimenter.
+
+### One Glean app ID per product domain
+
+In this option, we would have one Glean application ID that is shared across all applications within a given product domain.
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply docs/adrs/0011-nimbus-web-glean-application-structure.md:82]: Ideally, the Experimenter API would also expose this identifier/application since this could be used by Jetstream to generate more cost efficient queries.

**Comment by scholtzan** [comment]
I added some notes on the changes that would need to be done on the jetstream/mozanalysis side: https://docs.google.com/document/d/1kTwZ44o5EQaZhv6J9fpn09wpr4I4O8zI4Q9FAyLQgAw/edit

tl;dr we might want to make these changes anyway at some point to support experiments that run on multiple channels. Jetstream could rely on the (relatively new) UNIONED app name datasets, instead of the per app-ID datasets.

File: docs/adrs/0011-nimbus-web-glean-application-structure.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+In the previous example, the product domain is Pocket.
+
+The problem this ADR addresses is as follows: with Cirrus, how should we structure Glean applications for cross-platform apps integrating Cirrus?
+
+> #### Note
+> Various explanations below will reference Pocket, but this ADR generally applies to all products planning to integrate Nimbus into their applications.  
+
+## Decision Drivers
+
+* [Probe-scraper repository list](https://github.com/mozilla/probe-scraper/blob/main/repositories.yaml) developer burden
```
[the remark this answers is not in the document]
Comment (jaredlockhart) [reply docs/adrs/0011-nimbus-web-glean-application-structure.md:30]: Can you expand on this and explain in more detail what this means?

File: docs/adrs/0011-nimbus-web-glean-application-structure.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+With this option, we would have one Glean application ID that is shared across all instances of Cirrus.
+All applications using Cirrus would share a single Glean application ID, allowing for an overall lower maintenance burden for probe-scraper and easier collation of data across applications.
+One downside of this solution is that data from multiple product domains will exist within the same table structure.
+
+* Good, because it has the smallest footprint in the probe-scraper repository of the listed options.
+* Good, because the Glean SDK supports this particular use case.
+* Good, because this model supports Pocket's architecture.
+* Bad, because data retention cannot be handled separately for each application or product domain.
+* Bad, because it breaks away from our existing Glean usage practices.
```
Comment (jaredlockhart) [inline docs/adrs/0011-nimbus-web-glean-application-structure.md:74]: I'm not sure what this means?  How does this impact using Glean in our project?

File: docs/adrs/0011-nimbus-web-glean-application-structure.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+* Bad, because it requires changes to Jetstream.
+  * At this time, the nature and timeline of these changes is unclear.
+* Bad, because it could require more changes to Experimenter than we already expect.
+  * While we will still need to add new applications, it is also possible we will need to adjust how applications are handled in Experimenter.
```
Comment (jaredlockhart) [inline docs/adrs/0011-nimbus-web-glean-application-structure.md:78]: Yeah, now we start to hit the limits of conflating a telemetry identifier with a targeting identifier.

File: docs/adrs/0011-nimbus-web-glean-application-structure.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaredlockhart) [reply docs/adrs/0011-nimbus-web-glean-application-structure.md:82]: Interesting.  Then we've got two identifiers, a broader domain id, and a specific app id.  Both for telemetry and targeting.  Probably some nuance there to sort out.

**Review by jaredlockhart — APPROVED** [review]
Thanks for writing all this up @jeddai !  I agree with the decided solution.  We'll have to figure out what having the pair of `(domain id, app id)` means for telemetry and targeting, but that should be manageable.  Given the constraints you've identified this seems like the most approachable strategy.  I'm happy to approve this and then move forward with a plan for how to manage these ids in deploying cirrus, configuring its telemetry, and changes necessary for Experimenter and Jetstream.  👍

**Comment by badboy** [comment]
> 



> I added some notes on the changes that would need to be done on the jetstream/mozanalysis side: [docs.google.com/document/d/1kTwZ44o5EQaZhv6J9fpn09wpr4I4O8zI4Q9FAyLQgAw/edit](https://docs.google.com/document/d/1kTwZ44o5EQaZhv6J9fpn09wpr4I4O8zI4Q9FAyLQgAw/edit)
> 
> tl;dr we might want to make these changes anyway at some point to support experiments that run on multiple channels. Jetstream could rely on the (relatively new) UNIONED app name datasets, instead of the per app-ID datasets.

The individual changes don't seem to hard, though of course doing them all might require more work to ensure it works right.
As I have no knowledge about responsibilities between experimenter, jetstream and mozanalysis: Who would be the right person to tackle this?

File: docs/adrs/0011-nimbus-web-glean-application-structure.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeddai) [reply docs/adrs/0011-nimbus-web-glean-application-structure.md:30]: Travis had some comments about it and regrettably I no longer remember what they were. He's out for a few weeks for moving, so for now I'm going to remove this line. I'll follow up with him on it later and circle back here if needed.

File: docs/adrs/0011-nimbus-web-glean-application-structure.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeddai) [reply docs/adrs/0011-nimbus-web-glean-application-structure.md:82]: Yep, I'll add that as a note here and in the single-glean-app option

File: docs/adrs/0011-nimbus-web-glean-application-structure.md (hunk 3 above)
Comment (jeddai) [reply docs/adrs/0011-nimbus-web-glean-application-structure.md:74]: This is referring to how Glean applications are configured at this time — one app id per application (fenix, firefox_ios, etc) and its use in client-side applications versus servers.

This has no impact as far as I can tell, just wanted to note that this option is a departure from the norm.

**Comment by jeddai** [comment]
> As I have no knowledge about responsibilities between experimenter, jetstream and mozanalysis: Who would be the right person to tackle this?

That's certainly a great question, I'm not entirely sure either but I'd definitely lean towards coordinating closely (if only to help the Nimbus folks be more familiar with jetstream/mozanalysis).

**Comment by scholtzan** [comment]
> Who would be the right person to tackle this?

Probably either Mike or I
